### ISSUE TITLE
Potential infinite loop - waited 4+ minutes to minimize

### ISSUE THREAD

**Issue by jonathan-m-phillips** [kelloggm/specimin#152]
Dry running ASHE Automation, I waited over 4 minutes for this target method to minimize before ending the process.

--args=--outputDirectory "temp/dir" --root "/Users/jonathanphillips/Desktop/NJIT/checker-framework/dataflow/src/main/java" --targetFile "org/checkerframework/dataflow/analysis/AbstractAnalysis.java" --targetMethod "org.checkerframework.dataflow.analysis.AbstractAnalysis#getValue(Tree)"

**Comment by kelloggm** [comment]
Confirmed locally. I'm working on this.

**Comment by kelloggm** [comment]
I've isolated the problem in this case, but I'm not sure how to fix it. @LoiNguyenCS, please read through this comment and then let me know what you think.

This is caused by an infinite loop in `UnsolvedSymbolVisitor` when handling a method call expression of the form:
```
getValue(cfg.getTreeLookup().get(t))
```

`getValue` is the target method, so this needs to be resolved. For context, the full target method is:
```
  @Override
  public @Nullable V getValue(Tree t) {
    // Dataflow is analyzing the tree, so no value is available.
    if (t == currentTree || cfg == null) {
      return null;
    }
    V result = getValue(getNodesForTree(t));
    if (result == null) {
      result = getValue(cfg.getTreeLookup().get(t));
    }
    return result;
  }
```

However, resolution for this call is failing (i.e., `canBeSolved(method)` is false in `UnsolvedSymbolVisitor#visit(MethodCallExpr, Void)`, causing `gotException` to be set to true at the end of the method). However, no change is induced, so Specimin enters an infinite loop. Interestingly, however, the _parameters_ of the method _are_ solvable - it's just that they're wrong.

The reason that the call remains unsolvable is that `cfg.getTreeLookup().get(t)` (i.e., the parameter) is given the type `ReferenceType{org.plumelib.util.GetReturnType}` (with an empty type parameters map). Note that this type is synthetic. However, this synthetic type doesn't match the parameter type of `getValue`, which is `Tree`. The problem might be related to one of the following properties of this example:
* this method overrides another method from an interface (`Analysis`).
* there is another method with a different parameter type that's also named `getValue` in this class (and in the implemented interface); it's parameter type is `Node`. This particular failing call is a recursive call to the `Tree` version of `getValue`, but it's possible that the presence of this other method has broken some invariant of Specimin.

**Comment by kelloggm** [comment]
I would expect that `javac` should correct the `GetReturnType` type, so maybe that's where the bug is?

**Comment by LoiNguyenCS** [comment]
Professor,

I don't think `gotException` is set to true because we have type mismatches. JavaParser does not require type matching for a type to be solved, which is why we needed `javac` in the first place. And `javac` is called after `UnsolvedSymbolVisitor` has finished its execution, so I don't think this bug is related to `javac`.

**Comment by LoiNguyenCS** [comment]
I think we have a similar problem in cf-3012, where a method call remains unsolved even though all the related parameters are solvable. I'm still not sure what the reason might be.

**Comment by kelloggm** [comment]
> And javac is called after UnsolvedSymbolVisitor has finished its execution, so I don't think this bug is related to javac

Maybe the problem then is that we need to interleave the calls to UnsolvedSymbolVisitor and to javac? That is, if UnsolvedSymbolVisitor gets stuck (which we'd need to detect somehow), then we can try moving on to javac, and then return to UnsolvedSymbolVisitor later. Let's discuss this later today in our meeting.

**Comment by kelloggm** [comment]
Fixed by #166
